main/snprintf.c

```diff
@@ -143,8 +143,12 @@ PHPAPI char *php_gcvt(double value, int ndigit, char dec_point, char exponent, c
 {
 	char *digits, *dst, *src;
 	int i, decpt, sign;
+	int mode = ndigit >= 0 ? 2 : 0;
 
-	digits = zend_dtoa(value, 2, ndigit, &decpt, &sign, NULL);
+	if (mode == 0) {
+		ndigit = 17;
+	}
+	digits = zend_dtoa(value, mode, ndigit, &decpt, &sign, NULL);
```

Comment:
Not everything goes though php_gcvt. Did you check how precision=-1 will work for php_fcvt/php_ecvt (and how it should work)? I think the RFC should also be more clear on how this change interacts with both userland and internal printf functions.